rover_gui/src/QNavigation/QNavigation.cpp

```diff
@@ -1 +1,250 @@
+    _ui.webViewContainer->setMaximumSize(1200, 1000);
+
+    _webChannel = new QWebChannel(this);
+    _webChannel->registerObject(QStringLiteral("bridge"), this);
+    _ui.webViewContainer->page()->setWebChannel(_webChannel);
+
+    connect(_ui.webViewContainer, &QWebEngineView::loadFinished, this, &QNavigation::onWebViewLoadFinished);
+    connect(_ui.setGoalButton, &QPushButton::clicked, this, &QNavigation::onSetGoalClicked);
+    connect(_ui.calculatePathButton, &QPushButton::clicked, this, &QNavigation::onCalculatePathClicked);
+    connect(_ui.waypointList, &QListWidget::itemClicked, this, &QNavigation::onWaypointSelected);
+    connect(_ui.clearWaypointsButton, &QPushButton::clicked, this, &QNavigation::onClearWaypointsClicked);
+    connect(_ui.clearPathButton, &QPushButton::clicked, this, &QNavigation::onClearPathClicked);
+    connect(_ui.deleteWaypointButton, &QPushButton::clicked, this, &QNavigation::onDeleteWaypointClicked);
+
+    _gpsSub
+        = _node->create_subscription<rover_msgs::msg::Gps>("/rover/gps/position",
+                                                           1,
+                                                           std::bind(&QNavigation::onGpsMessage, this, std::placeholders::_1));
+}
+
+void QNavigation::onGpsMessage(const rover_msgs::msg::Gps::SharedPtr msg_)
+{
+    _currentLat = msg_->latitude;
+    _currentLon = msg_->longitude;
+    _currentHeading = msg_->heading;
+
+    QMetaObject::invokeMethod(this,
+                              "gpsCallback",
+                              Qt::QueuedConnection,
+                              Q_ARG(double, msg_->latitude),
+                              Q_ARG(double, msg_->longitude),
+                              Q_ARG(double, msg_->heading));
+}
+
+void QNavigation::onSetGoalClicked()
+{
+    if (_ui.inputName->text().isEmpty() || _ui.inputLatitude->text().isEmpty() || _ui.inputLongitude->text().isEmpty())
+    {
+        QMessageBox::warning(this, "Input Error", "Please enter waypoint name and coordinates.");
+        return;
+    }
+
+    double lat_ = _ui.inputLatitude->text().toDouble();
+    double lon_ = _ui.inputLongitude->text().toDouble();
+    QString name_ = _ui.inputName->text();
+
+    for (const auto& waypoint : _waypoints)
+    {
+        if (waypoint.name == name_)
+        {
+            QMessageBox::warning(this,
+                                 "Duplicate Name",
+                                 "A waypoint with this name already exists. Please choose a different name.");
+            return;
+        }
+    }
+
+    QString id_ = "waypoint_" + QUuid::createUuid().toString(QUuid::WithoutBraces);
+
+    addWaypointToList(name_, lat_, lon_, id_);
+
+    emit sendGoal(name_, lat_, lon_, id_);
+
+    _ui.inputName->clear();
+    _ui.inputLatitude->clear();
+    _ui.inputLongitude->clear();
+}
+
+void QNavigation::pathDistanceCalculated(double distanceMeters_)
+{
+    QString distanceText_;
+    if (distanceMeters_ >= 1000)
```

Comment:
Comparing double with int
```suggestion
    if (distanceMeters_ >= 1000.0)
```